src/audio/copier/copier.c

```diff
@@ -253,15 +267,20 @@ static int create_host(struct comp_dev *parent_dev, struct copier_data *cd,
 				   ipc4_gtw_host, IPC4_DIRECTION(dir));
 	if (!cd->converter[IPC4_COPIER_GATEWAY_PIN]) {
 		comp_err(parent_dev, "failed to get converter for host, dir %d", dir);
-		ret = -EINVAL;
-		goto e_buf;
+		goto e_data;
 	}
 
 	cd->endpoint[cd->endpoint_num++] = dev;
+	cd->hd = comp_get_drvdata(dev);
 
-	return 0;
+	return ret;
 
+e_data:
+	rfree(hd);
+e_dev:
+	rfree(dev);
 e_buf:
+	ret = -EINVAL;
```

Comment:
this needs to be revmoed, The ret value should come from wherever the failure occurs

Reply:
ok, this is reasonable, just this will bring code looks like a little duplicated, I will change.